src/mem/cache/prefetch/bop.cc

```diff
@@ -114,22 +111,24 @@ BOP::delayQueueEventWrapper()
 }
 
 unsigned int
-BOP::hash(Addr addr, unsigned int way) const
+BOP::index(Addr addr, unsigned int way) const
 {
-    Addr hash1 = addr >> way;
-    Addr hash2 = hash1 >> floorLog2(rrEntries);
-    return (hash1 ^ hash2) & (Addr)(rrEntries - 1);
+    Addr lrrEntries = floorLog2(rrEntries);
+    Addr lineaddr = addr >> lBlkSize;
+    Addr hash = lineaddr ^ (lineaddr >> (lrrEntries << way));
```

Comment:
¨The RR table is accessed through a simple hash function.
For instance, for a 256-entry RR table, we XOR the 8 least
significant line address bits with the next 8 bits to obtain the
table index. For 12-bit tags, we skip the 8 least significant
line address bits and extract the next 12 bits.¨

- Here we use (64-8)-bit tags, right?
- The paper does not mention using the way here, and I don´t recall what is its used for. Could you document why?

Reply:
Since the current version of BOP was not working as expected, i.e. it wasn't providing the expected performance on Spec workloads, I started looking for other implementations of BOP to see what could be the problem. While examining the prefetcher, I found that it was experiencing a lot of hash collisions while indexing the RR Table. To figure out if this was intended behavior or not, I started searching for other implementations of BOP online and I found the original code submitted by the author for DPC-2 [here](https://comparch-conf.gatech.edu/dpc2/final_program.html).

If you look at the source code, you'll two functions `rr_index_left` and `rr_index_right` which are used to generate the index which would be used to index the RR table. This logic is implemented in the `index` function in this PR. In the author's code you'll see that he multiplies the RRINDEX by 2 for indexing the right side, whereas he doesn't do the same for the left side. In this PR the `index` function is called with 0 for left, whereas it is called with 1 for right  (refer to the enum `RRWay` defined at line 57 of `bop.hh`). Therefore, if we right shift with the parameter `way`, then when `way` corresponds to left, we'll right shift with 0, and when it corresponds to right, we'll right shift by 1, i.e. multiply by 2. Thus, this code is equivalent to what the author submitted for DPC-2.

Reply:
>> Here we use (64-8)-bit tags, right?
We will use 8 bit tags if there are 256 entries in the RR Table. While this is the default size of the RR Table, it is not a constant. One can in fact change the size if they wish to do so.

>> The paper does not mention using the way here, and I don´t recall what is its used for. Could you document why?
Since the reason why I made this change is not mentioned in the paper, should I still add a comment explaining the logic?

Reply:
Yes, please add a comment explaining this logic! Your explanation was great, and will make it much easier to understand the code.

Reply:
I've added a comment explaining the calculation